### ISSUE TITLE
Running an executable stalls

### ISSUE THREAD

**Issue by jkiviluo** [Spine-project/Spine-Toolbox#1344]
I tried to run parallel scenarios for a Windows executable with a considerable amount of input files. It stalls and does not seem to ever finish. Also the execution choice is messed up - shows up only one run even when there are two. Maybe due to lot of file names that the executable row tries to show.

You can try it out with this repository: https://gitlab.vtt.fi/FlexTool/flextool3/-/tree/master (access limited)
You need to populate the DB first and run the exportToCSV to get the files.

**Comment by PekkaSavolainen** [comment]
It seems to run ok, everything passes but It doesn't seem to run anything in parallel. Should I do some step myself that is not in the repo?

ImportOrig seems to have errors though, are these there on purpose?

```
[29-03-2021 14:55:25] ***Executing Importer ImportOrig***
[29-03-2021 14:55:27] Importing template.xlsm
[29-03-2021 14:55:40] Successfully read 922327 data.
[29-03-2021 14:55:42] Errors while importing a table.
[29-03-2021 14:55:42] Ignoring errors. Set Cancel import on error to bail out instead.
[29-03-2021 14:55:42] Inserted 629 data with 91 errors into sqlite:///C:\Users\ttepsa\OneDrive - Teknologian Tutkimuskeskus VTT\Documents\GIT\FLEXTOOL3\FlexTool2D_data.sqlite?spinedbfilter=cfg%3Aexecution%3A%7B%22execution_item%22%3A+%22ImportOrig%22%2C+%22scenarios%22%3A+%5B%5D%2C+%22timestamp%22%3A+%222021-03-29T14%3A55%3A25%22%7D
[29-03-2021 14:55:42] Error log
[29-03-2021 14:55:42] Executing Importer ImportOrig finished
```

**Comment by jkiviluo** [comment]
I realize that the import process don't include scenarios. I had that directly setup in the DB, which is not part of the repo. I guess you can just make another scenario, add the Base alternative there and run two scenarios (from the arrow).

Those import errors (empty values) are reported in another issue elsewhere assigned to Manuel.

**Comment by PekkaSavolainen** [comment]
By "execution choice" do you mean this red triangle

![issue_1344_workflow](https://user-images.githubusercontent.com/43600314/112987664-a757a800-916b-11eb-8154-64c864399eb7.png)

It does not show anything in the Link properties because running a Tool in parallel after an Exporter does not seem to be supported. Or am I missing something @manuelma, @soininen? Or if you (@jkiviluo) mean the other execution choice (dockwidget) that shows up sometimes, it only has one entry (the extremely long string) basically because of the same reason. Executing the Tool fails at least because it's trying to create a folder named by that extremely long string.

The Exporter in this dag creates two output folders when there are two scenarios coming from the Data Transformer. My idea is that we could add these folders under a new category into Link properties (ResourceFilterModel) and let the user select them. If the user selects them both and next one up is a Tool, then the Tool is run parallel.

There were other things I wanted to add but I cannot run the project anymore because of this

```
Traceback (most recent call last):
  File "C:\Users\ttepsa\OneDrive - Teknologian Tutkimuskeskus VTT\Documents\GIT\SPINEITEMS\spine_items\exporter\widgets\specification_editor_window.py", line 508, in _change_mapping_type
    self._undo_stack.push(SetExportObjectsFlag(index, mapping_specification.export_objects_flag))
AttributeError: 'MappingSpecification' object has no attribute 'export_objects_flag'
```

and also other tracebacks related to the same 'export_objects_flag' attribute.

**Comment by soininen** [comment]
> It does not show anything in the Link properties because running a Tool in parallel after an Exporter does not seem to be supported. Or am I missing something @manuelma, @soininen?

Running Tool parallel is supported and should work fine. However, there is nothing to edit in the link since Exporter does not supply database URLs and therefore does not support scenario/tool filters. You select those filters using the arrow between "Params" and "ExportToCSV" items.

**Comment by PekkaSavolainen** [comment]
>  However, there is nothing to edit in the link since Exporter does not supply database URLs and therefore does not support scenario/tool filters.

I noticed, and that's why I wrote that it does not seem to be supported.

I did select two scenarios from the arrow between "Params" and "ExportToCSV" items. But, I'm confused, how does FlexTool know that it's supposed to run in parallel in this DAG. The project items are only supposed to see the resources of their immediate neighbors.

**Comment by soininen** [comment]
> how does FlexTool know that it's supposed to run in parallel in this DAG?

Engine gives a hint to Tool's executable item, see `filter_id` in `ExecutableItemBase`. In case of Tool, it stores its output files into separate directories so that files from different 'parallel branches' do not mix. The script run by FlexTool can be completely oblivious of parallelism (we should fix #1243, though).

**Comment by PekkaSavolainen** [comment]
Thanks @soininen for the tip. I did notice that FlexTool tries to run in parallel but there's something not quite working (hence the issue). I'll check if it's related to #1243.

**Comment by jkiviluo** [comment]
This is still an issue. The project has been updated and I'm currently running the lower workflow. It stalls if I try to use scenarios to run. https://github.com/Spine-project/Spine-Toolbox/issues/1344

**Comment by jkiviluo** [comment]
It looks like it finishes the first scenario but does not start the second. It just keeps churning and uses a lot of system resources. Until I kill Spine Toolbox.

**Comment by PekkaSavolainen** [comment]
The problem here is the filter_id that grows too long with multiple input files. We use this filter_id string to create a directory. The character limit for a full path on Windows is 260 characters and on Linux 4096 characters. I'm not sure how to fix that because I don't understand the reasoning for creating such long paths in the first place. @manuelma any improvement ideas here?

**Comment by jkiviluo** [comment]
The scenario name with a possible time_stamp should be sufficient to distinguishes scenario runs.

**Comment by manuelma** [comment]
Scenario name is not enough. Multiple resources can be coming into the tool from multiple items, and each of these resources might be filtered by a combination of scenario and tool. So it's resource names, plus scenario, plus tool.

**Comment by PekkaSavolainen** [comment]
But the time_stamp still distinguishes the scenario runs so there's no overlap. Would it make sense to make some simple metadata text file into the result folder to add this additional info there to limit the folder name length.

**Comment by soininen** [comment]
Time stamps are nasty because you might create two with the same value. I suggest we use some hash algorithm instead.

**Comment by manuelma** [comment]
I think the solution is the combination of the two comments above. So the filter id can be just the hash, plus the scenario, plus the tool (don't forget the tool, it's also a filter).

And within the folder, we can have a 'dot execution details' text or json file with all the detailed information about resource, scenario, and tool.

**Comment by jkiviluo** [comment]
Bumping this, since I'm still bumping into this when trying to run parallel.

**Comment by timok-vtt** [comment]
There might be something else here as well than just the path name. I changed the path name to  be just a hash of the filter id. Now it runs on one scenario, but selecting multiple scenarios will still cause Spine Toolbox to hang.  

Selecting just one will run when making sure the path names don't clash and the directory can actually be created.
